// react-screenshots/lib/Screenshots/ScreenshotsMagnifier/index.js
import { jsx, jsxs } from "react/jsx-runtime";
import { memo, useEffect, useLayoutEffect, useRef, useState } from "react";
import useLang from "../hooks/useLang.js";
import useStore from "../hooks/useStore.js";
import "./index.css";
const magnifierWidth = 100;
const magnifierHeight = 80;
const Screenshots_ScreenshotsMagnifier = /*#__PURE__*/ memo(function({ x, y }) {
    const { width, height, image } = useStore();
    const lang = useLang();
    const [position, setPosition] = useState(null);
    const elRef = useRef(null);
    const canvasRef = useRef(null);
    const ctxRef = useRef(null);
    const [rgb, setRgb] = useState('000000');
    useLayoutEffect(()=>{
        if (!elRef.current) return;
        const elRect = elRef.current.getBoundingClientRect();
        let tx = x + 20;
        let ty = y + 20;
        if (tx + elRect.width > width) tx = x - elRect.width - 20;
        if (ty + elRect.height > height) ty = y - elRect.height - 20;
        if (tx < 0) tx = 0;
        if (ty < 0) ty = 0;
        setPosition({
            x: tx,
            y: ty
        });
    }, [
        width,
        height,
        x,
        y
    ]);
    useEffect(()=>{
        if (!image || !canvasRef.current) {
            ctxRef.current = null;
            return;
        }
        if (!ctxRef.current) ctxRef.current = canvasRef.current.getContext('2d');
        if (!ctxRef.current) return;
        const ctx = ctxRef.current;
        ctx.clearRect(0, 0, magnifierWidth, magnifierHeight);
        const rx = image.naturalWidth / width;
        const ry = image.naturalHeight / height;
        ctx.drawImage(image, x * rx - magnifierWidth / 2, y * ry - magnifierHeight / 2, magnifierWidth, magnifierHeight, 0, 0, magnifierWidth, magnifierHeight);
        const { data } = ctx.getImageData(Math.floor(magnifierWidth / 2), Math.floor(magnifierHeight / 2), 1, 1);
        const hex = Array.from(data.slice(0, 3)).map((val)=>val >= 16 ? val.toString(16) : `0${val.toString(16)}`).join('').toUpperCase();
        setRgb(hex);
    }, [
        width,
        height,
        image,
        x,
        y
    ]);
    return /*#__PURE__*/ jsxs("div", {
        ref: elRef,
        className: "screenshots-magnifier",
        style: {
            transform: `translate(${position?.x}px, ${position?.y}px)`
        },
        children: [
            /*#__PURE__*/ jsx("div", {
                className: "screenshots-magnifier-body",
                children: /*#__PURE__*/ jsx("canvas", {
                    ref: canvasRef,
                    className: "screenshots-magnifier-body-canvas",
                    width: magnifierWidth,
                    height: magnifierHeight
                })
            }),
            /*#__PURE__*/ jsxs("div", {
                className: "screenshots-magnifier-footer",
                children: [
                    /*#__PURE__*/ jsxs("div", {
                        className: "screenshots-magnifier-footer-item",
                        children: [
                            lang.magnifier_position_label,
                            ": (",
                            x,
                            ",",
                            y,
                            ")"
                        ]
                    }),
                    /*#__PURE__*/ jsxs("div", {
                        className: "screenshots-magnifier-footer-item",
                        children: [
                            "RGB: #",
                            rgb
                        ]
                    })
                ]
            })
        ]
    });
});
export { Screenshots_ScreenshotsMagnifier as default };
